design/crd-configstate.md

```diff
@@ -0,0 +1,283 @@
+	// Empty when Result is "Valid".
+	// +optional
+	LastError string `json:"lastError,omitempty"`
+}
+
+```
+
+* Initially, users will query individual ConfigurationState CRs per component
+* Aggregation can be addressed in a future proposal/issue if needed
+* Users can leverage label selectors to query specific subsets (e.g., all speakers, specific node)
+* API does not provide the summary of the "input configuration", only the error
+  which should clue the user which CR must be changed
+
+## Controller side
+
+For Controller (Deployment - single instance) there are three reconcilers
+
+- PoolReconciler - yes but only config error, that error should appear in configState
+- PoolStatusReconciler - no config errors
+- ServiceReconciler - no config errors
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: ConfigurationState
+metadata:
+    name: controller
+    namespace: metallb-system
+    labels:
+        metallb.io/component-type: controller
+status:
+    result: "Valid"
+    lastError: ""
+
+# Failed example
+apiVersion: metallb.io/v1beta1
+kind: ConfigurationState
+metadata:
+    name: controller
+    namespace: metallb-system
+    labels:
+        metallb.io/component-type: controller
+status:
+    result: "Invalid"
+    lastError: "failed to parse configuration: CIDR \"192.168.10.100/32\" in pool \"client2-pool\" overlaps with already defined CIDR \"192.168.10.0/24\""
+```
+
+## Speaker side
+
+For Speaker (DaemonSet - one pod per node) there are
+
+- ConfigReconciler - yes, but only config error, that error should appear in configState
+- NodeReconciler - (tentative yes, under investigation //TODO)
+- FRRK8sReconciler - no config errors
+- ServiceReconciler - no config errors
+- ServiceBGPStatusReconciler - no config errors
+- Layer2StatusReconciler - no config errors
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: ConfigurationState
+metadata:
+    name: speaker-kind-worker
+    namespace: metallb-system
+    labels:
+        metallb.io/component-type: speaker
+        metallb.io/node-name: kind-worker
+status:
+    result: "Valid"
+    lastError: ""
+
+# Failed example:
+apiVersion: metallb.io/v1beta1
+kind: ConfigurationState
+metadata:
+    name: speaker-kind-worker2
+    namespace: metallb-system
+    labels:
+        metallb.io/component-type: speaker
+        metallb.io/node-name: kind-worker2
+status:
+    result: "Invalid"
+    lastError: "failed to parse configuration: invalid BGP peer address"
+```
+
+## Implementation
+
+Ideally like FRRNodeState pattern it metallb/frrk8s repo, but might not be
+ideal because for ConfigurationStateReconciler needs references to ConfigReconciler
+and NodeReconciler in-memory result. Alternative implementation based on
+condition instead of channel can be evaluated. TBD and discussed during
+the implementation PR.
+
+`res := r.Handler(r.Logger, cfg)` might need refactor to return error.
+
+**Note:** Subset of E2E test will work only with disabling the webhook
```

Comment:
given the transient error considerations, I think we can test without disabling the webhooks, right?

Reply:
for the transient errors, correct no webhook disable is needed (and probably we are ok)